File: frontend/interviewee-app/scripts/generate-icons.js
#!/usr/bin/env node
/**
 * Icon Generation Script for Blockd Interviewee Electron App
 *
 * This script converts a source PNG image into all required icon formats
 * for Windows (.ico), macOS (.icns), and Linux (.png).
 *
 * Usage:
 *   npm run generate-icons
 *   OR
 *   node scripts/generate-icons.js [source-png-path]
 *
 * Requirements:
 *   - electron-icon-builder package (installed as devDependency)
 *   - Source PNG should be at least 1024x1024 pixels (square)
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// Configuration
const RESOURCES_DIR = path.join(__dirname, '..', 'resources');
const DEFAULT_SOURCE = 'Electron_Logo.png';

// Get source file from argument or use default
const sourceArg = process.argv[2];
let sourcePng;

if (sourceArg) {
  sourcePng = path.isAbsolute(sourceArg) ? sourceArg : path.join(process.cwd(), sourceArg);
} else {
  // Look for source in resources directory
  const possibleNames = ['Electron_Logo.png', 'logo.png', 'icon-source.png', 'icon.png'];
  for (const name of possibleNames) {
    const testPath = path.join(RESOURCES_DIR, name);
    if (fs.existsSync(testPath)) {
      sourcePng = testPath;
      break;
    }
  }
}

if (!sourcePng || !fs.existsSync(sourcePng)) {
  console.error('Error: Source PNG file not found!');
  console.error('');
  console.error('Please ensure your logo PNG file exists in one of these locations:');
  console.error(`  - ${path.join(RESOURCES_DIR, 'Electron_Logo.png')}`);
  console.error(`  - ${path.join(RESOURCES_DIR, 'logo.png')}`);
  console.error(`  - ${path.join(RESOURCES_DIR, 'icon-source.png')}`);
  console.error('');
  console.error('Or provide the path as an argument:');
  console.error('  npm run generate-icons -- /path/to/your/logo.png');
  process.exit(1);
}

console.log('='.repeat(60));
console.log('Blockd Icon Generator');
console.log('='.repeat(60));
console.log(`Source: ${sourcePng}`);
console.log(`Output: ${RESOURCES_DIR}`);
console.log('');

// Ensure resources directory exists
if (!fs.existsSync(RESOURCES_DIR)) {
  fs.mkdirSync(RESOURCES_DIR, { recursive: true });
}

try {
  // Run electron-icon-builder
  console.log('Generating icons...');

  const iconBuilderPath = path.join(__dirname, '..', 'node_modules', '.bin', 'electron-icon-builder');
  const cmd = `"${iconBuilderPath}" --input="${sourcePng}" --output="${RESOURCES_DIR}" --flatten`;

  execSync(cmd, { stdio: 'inherit' });

  // Rename output files to match forge.config.ts expectations
  const generatedFiles = {
    mac: path.join(RESOURCES_DIR, 'icon.icns'),
    win: path.join(RESOURCES_DIR, 'icon.ico'),
    png: path.join(RESOURCES_DIR, 'icon.png'),
  };

  // Copy the 1024x1024 PNG as icon.png if it doesn't exist
  const png1024 = path.join(RESOURCES_DIR, '1024x1024.png');
  if (fs.existsSync(png1024) && !fs.existsSync(generatedFiles.png)) {
    fs.copyFileSync(png1024, generatedFiles.png);
  }

  console.log('');
  console.log('='.repeat(60));
  console.log('Icons generated successfully!');
  console.log('='.repeat(60));
  console.log('');
  console.log('Generated files:');

  for (const [platform, filePath] of Object.entries(generatedFiles)) {
    if (fs.existsSync(filePath)) {
      const stats = fs.statSync(filePath);
      console.log(`  [${platform.toUpperCase().padEnd(5)}] ${path.basename(filePath)} (${(stats.size / 1024).toFixed(1)} KB)`);
    }
  }

  console.log('');
  console.log('Your Electron app is now ready with custom icons!');
  console.log('Run "npm run make" to build the application.');

} catch (error) {
  console.error('');
  console.error('Error generating icons:', error.message);
  console.error('');
  console.error('Make sure electron-icon-builder is installed:');
  console.error('  npm install --save-dev electron-icon-builder');
  process.exit(1);
}

File: chromium/src/chrome/app/theme/blocked/generate_icons.py
#!/usr/bin/env python3
"""
Generate app icons in all required sizes for Chromium browser branding.

This script generates PNG icons from a source image in all sizes required for:
- Windows: 16, 32, 48, 256
- macOS: 16, 32, 128, 256, 512, 1024
- Linux: 16, 32, 48, 64, 128, 256

Usage:
    python generate_icons.py [source_image]

If no source image is provided, defaults to logo_source.png in the same directory.

Requirements:
    - Pillow (pip install Pillow)
"""

import subprocess
import sys
from pathlib import Path


def main():
    script_dir = Path(__file__).parent
    source_image = script_dir / "logo_source.png"

    if len(sys.argv) > 1:
        source_image = Path(sys.argv[1])

    if not source_image.exists():
        print(f"Error: Source image not found: {source_image}")
        sys.exit(1)

    # All required sizes (union of Windows, macOS, and Linux requirements)
    sizes = [16, 32, 48, 64, 128, 256, 512, 1024]

    print(f"Generating icons from: {source_image}")

    for size in sizes:
        output_file = script_dir / f"product_logo_{size}.png"

        try:
            # Use macOS sips command to resize
            subprocess.run([
                "sips",
                "-z", str(size), str(size),  # Resize to size x size
                str(source_image),
                "--out", str(output_file)
            ], check=True, capture_output=True)

            print(f"  Created: product_logo_{size}.png")
        except subprocess.CalledProcessError as e:
            print(f"  Error creating {size}x{size}: {e.stderr.decode()}")
        except FileNotFoundError:
            print("  Error: sips command not found. This script requires macOS.")
            print("  On other platforms, use ImageMagick or Pillow directly.")
            sys.exit(1)

    print("\nIcon generation complete!")
    print("\nNext steps:")
    print("  1. Review generated icons for quality")
    print("  2. Run generate_ico.py to create Windows .ico file")
    print("  3. Run generate_icns.py to create macOS .icns file")


if __name__ == "__main__":
    main()
